internal/gitprovider/azure/azure.go

```diff
@@ -0,0 +1,234 @@
+package azure
+
+import (
+	"context"
+	"fmt"
+	"net/url"
+	"slices"
+	"strings"
+
+	"github.com/akuity/kargo/internal/git"
+	"github.com/akuity/kargo/internal/gitprovider"
+	"k8s.io/utils/ptr"
+
+	"github.com/microsoft/azure-devops-go-api/azuredevops/v7"
+	adogit "github.com/microsoft/azure-devops-go-api/azuredevops/v7/git"
+)
+
+const ProviderName = "azure"
+
+// Azure DevOps URLs can be of two different forms:
+//
+//   - https://dev.azure.com/org/project/_git/repo
+//   - https://org.visualstudio.com/project/_git/repo
+//
+// We support both forms.
+var providerSuffixes = []string{"dev.azure.com", "visualstudio.com"}
+
+var registration = gitprovider.Registration{
+	Predicate: func(repoURL string) bool {
+		u, err := url.Parse(repoURL)
+		if err != nil {
+			return false
+		}
+		return slices.ContainsFunc(providerSuffixes, func(suffix string) bool {
+			return strings.HasSuffix(u.Host, suffix)
+		})
+	},
+	NewProvider: func(
+		repoURL string,
+		opts *gitprovider.Options,
+	) (gitprovider.Interface, error) {
+		return NewProvider(repoURL, opts)
+	},
+}
+
+func init() {
+	gitprovider.Register(ProviderName, registration)
+}
+
+type provider struct {
+	org        string
+	project    string
+	repo       string
+	connection *azuredevops.Connection
+}
+
+// NewProvider returns an Azure DevOps-based implementation of gitprovider.Interface.
+func NewProvider(
+	repoURL string,
+	opts *gitprovider.Options,
+) (gitprovider.Interface, error) {
+	if opts == nil || opts.Token == "" {
+		return nil, fmt.Errorf("options are required for Azure DevOps provider")
+	}
+	org, project, repo, err := parseRepoURL(repoURL)
+	if err != nil {
+		return nil, fmt.Errorf("error creating Azure DevOps provider: %w", err)
+	}
+	organizationUrl := fmt.Sprintf("https://dev.azure.com/%s", org)
+	connection := azuredevops.NewPatConnection(organizationUrl, opts.Token)
+
+	return &provider{
+		org:        org,
+		project:    project,
+		repo:       repo,
+		connection: connection,
+	}, nil
+}
+
+// CreatePullRequest implements gitprovider.Interface.
+func (p *provider) CreatePullRequest(
+	ctx context.Context,
+	opts *gitprovider.CreatePullRequestOpts,
+) (*gitprovider.PullRequest, error) {
+	gitClient, err := adogit.NewClient(ctx, p.connection)
+	if err != nil {
+		return nil, err
```

Comment:
Nit: I'd wrap this one to indicate the problem was getting an ado client.

Reply:
done